HiAnalysis/HiOnia/plugins/HiOniaAnalyzer.cc

```diff
@@ -1587,31 +2161,220 @@ HiOniaAnalyzer::fillGenInfo()
+	      }
+	      else if ( isNeutrino(genDau2->pdgId()) && (abs(genDau3->pdgId()) == 13) ){
+		reco::GenParticleRef gennuW = genDau2;
+		reco::GenParticleRef genmuW = genDau3;
+	      }
+	      else if ( isNeutrino(genDau3->pdgId()) && (abs(genDau2->pdgId()) == 13) ){
+		reco::GenParticleRef gennuW = genDau3;
+		reco::GenParticleRef genmuW = genDau2;
+	      }
+	      else {
+		goodDaughters = false;
+	      }
+	    
+	      //Fill info for Bc and its mu,nu daughters
+	      // if (verbose){
+	      //   std::cout<<"BcPID = "<<genBc->pdgId()<<std::endl;
+	      //   std::cout<<"Bc 1st daughter PID = "<<genDau1->pdgId()<<std::endl;
+	      //   std::cout<<"Bc 2nd daughter PID = "<<genDau2->pdgId()<<std::endl;
+	      //   std::cout<<"Bc 3rd daughter PID = "<<genDau3->pdgId()<<std::endl;
+	      //   std::cout<<"Jpsi PID = "<<gen->pdgId()<<std::endl;
+	      //   std::cout<<"status of neutrino daughter of Bc = "<<gennuW->status()<<std::endl;
+	      //   std::cout<<"status of muon daughter of Bc = "<<genmuW->status()<<std::endl;
+	      // }
+	      if(goodDaughters && (genmuW->charge() == genBc->charge()) 
+		 && ( genmuW->status() == 1 ) ){
+	
+		Gen_QQ_Bc_idx[Gen_QQ_size] = Gen_Bc_size;
+		Gen_Bc_QQ_idx[Gen_Bc_size] = Gen_QQ_size;
+
+		Gen_Bc_pdgId[Gen_Bc_size] = genBc->pdgId();
+		std::pair<int, std::pair<float, float> > MCinfo = findGenBcInfo(genBc, gen);
+		Gen_Bc_ctau[Gen_Bc_size] = 10.0*MCinfo.second.first;
+
+		TLorentzVector vBc = lorentzMomentum(genBc->p4());
+		new((*Gen_Bc_4mom)[Gen_Bc_size])TLorentzVector(vBc);
+
+		TLorentzVector vmuW = lorentzMomentum(genmuW->p4());
+		Gen_Bc_muW_idx[Gen_Bc_size] = IndexOfThisMuon(&vmuW , Gen_mu_4mom, true);
+	      
+		TLorentzVector vnuW = lorentzMomentum(gennuW->p4());
+		new((*Gen_Bc_nuW_4mom)[Gen_Bc_size])TLorentzVector(vnuW);
+	    
+		Gen_Bc_size++;
+	      }
+	      else {std::cout<<"WARNING : Problem with daughters of the Bc, hence Bc and its daughters are not written out"<<std::endl;}
+	    }
+	  }
           Gen_QQ_size++;
         }
       }
+    }
+  }  
+  return;
+}
 
-      if (abs(gen->pdgId()) == 13  && gen->status() == 1) {
-        Gen_mu_type[Gen_mu_size] = _isPromptMC ? 0 : 1; // prompt: 0, non-prompt: 1
-        Gen_mu_charge[Gen_mu_size] = gen->charge();
-
-        TLorentzVector vMuon = lorentzMomentum(gen->p4());
-        new((*Gen_mu_4mom)[Gen_mu_size])TLorentzVector(vMuon);
+//Find the indices of the reconstructed muon matching each generated muon, and vice versa
+void
+HiOniaAnalyzer::fillMuMatchingInfo()
+{
+ 
+  //Find the index of generated muon associated to a reco muon, txs to Reco_mu_pTrue
+  for (int irec=0;irec<Reco_mu_size;irec++){
+    int foundGen = -1;
+    if(Reco_mu_pTrue[irec]>=0){ //if pTrue=-1, then the reco muon is a fake                                                                                                                                                                  
+
+      for (int igen=0;igen<Gen_mu_size;igen++){
```

Comment:
I think it would be wrong because this is not what happens in data : you could have a true mu- reconstructed as a mu-, and we don't want to throw these out for MC, right?

Reply:
Well, at the PAT level, the MC matching is already done requiring same charge between reco and gen muon (check PhysicsTools/PatAlgos/python/mcMatchLayer0/muonMatch_cfi.py#L12).
But fine, lets not impose a charge condition on the Reco-MC matching at onia analyzer